File: solutions/typescript/157-middle-of-the-linked-list.ts
/**
 * Problem 157: Middle of the Linked List (LeetCode 876)
 * Difficulty: Easy
 * Language: TypeScript
 */
interface ListNode { val: number; next: ListNode | null; }
function middleNode(head: ListNode | null): ListNode | null {
  const nodes = (function collect(n: ListNode | null): ListNode[] {
    return n === null ? [] : [n, ...collect(n.next)];
  })(head);
  return nodes[Math.floor(nodes.length / 2)] ?? null;
}

File: solutions/rust/157-middle-of-the-linked-list.rs
/**
 * Problem 157: Middle of the Linked List (LeetCode 876)
 * Difficulty: Easy
 * Language: Rust
 */
#[derive(Clone)]
struct ListNode { val: i32, next: Option<Box<ListNode>> }
fn middle_node(head: &Option<Box<ListNode>>) -> Option<&ListNode> {
    let nodes: Vec<&ListNode> = std::iter::successors(head.as_deref(), |n| n.next.as_deref()).collect();
    nodes.get(nodes.len() / 2).copied()
}
